src/crimson/osd/heartbeat.cc

```diff
@@ -303,40 +303,40 @@ seastar::future<> Heartbeat::maybe_share_osdmap(
   Ref<MOSDPing> m)
 {
   const osd_id_t from = m->get_source().num();
-  const epoch_t osdmap_epoch = service.get_map()->get_epoch();
-  const epoch_t peer_epoch = m->map_epoch;
+  const epoch_t current_osdmap_epoch = service.get_map()->get_epoch();
   auto found = peers.find(from);
   if (found == peers.end()) {
     return seastar::now();
   }
   auto& peer = found->second;
 
-  if (peer_epoch > peer.get_last_epoch_sent()) {
-    logger().debug("{} updating session's last epoch sent "
-                   "from {} to peer's (id: {}) map epoch of {}",
-                   __func__, peer.get_last_epoch_sent(),
-                   from, peer_epoch);
-    peer.set_last_epoch_sent(peer_epoch);
+  // Peer already has newer projeceted osdmap epoch,
+  // no new incremental maps to share.
+  if (current_osdmap_epoch <= peer.get_projected_epoch()) {
+    logger().debug("{} our osdmap epoch of {} is not newer "
+                   " than peer {} projected_epoch of {}",
+                   __func__, current_osdmap_epoch, from,
+                   peer.get_projected_epoch());
+    return seastar::now();
   }
 
-  if (osdmap_epoch <= peer.get_last_epoch_sent()) {
-    logger().info("{} latest epoch sent {} is already later "
-                  "than osdmap epoch of {}",
-                  __func__ , peer.get_last_epoch_sent(),
-                  osdmap_epoch);
+  if (m->map_epoch <= peer.get_projected_epoch()) {
+    logger().debug("{} message's epoch of {} is not newer "
+                   " than peer {} projected_epoch of {}",
+                   __func__, m->map_epoch, from,
+                   peer.get_projected_epoch());
     return seastar::now();
   }
 
```

Comment:
why do you remove if  m->map_epoch > peer.get_projected_epoch to update session.set_projected_epoch?  in classic logic, even m->map_epoch <= peer.get_projected_epoch() still need send new epoch。

Reply:
Reverted.